File: backend/app/schemas/finding.py
import uuid
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


class SecurityRuleResponse(BaseModel):
    id: uuid.UUID
    rule_id: str
    title: str
    description: str
    service: str
    resource_type: str
    severity: str
    category: str
    remediation: str
    references: List[str]
    enabled: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RuleListResponse(BaseModel):
    items: List[SecurityRuleResponse]
    total: int
    page: int
    limit: int


class FindingResponse(BaseModel):
    id: uuid.UUID
    rule_id: uuid.UUID
    scan_id: uuid.UUID
    cloud_account_id: uuid.UUID
    resource_id: uuid.UUID
    finding_identifier: str
    title: str
    description: str
    severity: str
    risk_score: float
    risk_level: str = "MEDIUM"
    risk_priority: str = "MEDIUM"
    risk_factors: Dict[str, Any] = Field(default_factory=dict)
    risk_explanation: Optional[str] = None
    risk_calculated_at: Optional[datetime] = None
    status: str
    remediation: Optional[str] = None
    first_detected: datetime
    last_detected: datetime
    resolved_at: Optional[datetime] = None
    created_at: datetime

    # Denormalized context fields for UI convenience
    rule_code: Optional[str] = None
    service: Optional[str] = None
    resource_name: Optional[str] = None
    resource_identifier: Optional[str] = None

    model_config = {"from_attributes": True}


class FindingDetailResponse(FindingResponse):
    evidence: Dict[str, Any]
    references: Optional[List[str]] = None


class FindingListResponse(BaseModel):
    items: List[FindingResponse]
    total: int
    page: int
    limit: int


class FindingStatusUpdate(BaseModel):
    status: str = Field(..., description="Target status: OPEN, IN_PROGRESS, RESOLVED, ACCEPTED_RISK, FALSE_POSITIVE")
    rationale: Optional[str] = Field(None, max_length=1000, description="Justification or context for the status update")


class FindingNoteCreate(BaseModel):
    note: str = Field(..., min_length=1, max_length=5000, description="Analyst note or investigation details")


class FindingNoteResponse(BaseModel):
    id: uuid.UUID
    finding_id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    author_name: Optional[str] = "System Analyst"
    note: str
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/finding.ts
export interface SecurityRule {
  id: string;
  rule_id: string;
  title: string;
  description: string;
  service: string;
  resource_type: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  category: string;
  remediation: string;
  references: string[];
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface RuleListResponse {
  items: SecurityRule[];
  total: number;
  page: number;
  limit: number;
}

export interface Finding {
  id: string;
  rule_id: string;
  scan_id: string;
  cloud_account_id: string;
  resource_id: string;
  finding_identifier: string;
  title: string;
  description: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  risk_score: number;
  risk_level: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'INFO';
  risk_priority: 'IMMEDIATE' | 'HIGH' | 'MEDIUM' | 'LOW';
  risk_factors?: {
    severity: number;
    exposure: number;
    asset_criticality: number;
    exploitability: number;
    data_sensitivity: number;
    config_weakness: number;
    reasons?: Record<string, string>;
  };
  risk_explanation?: string;
  risk_calculated_at?: string;
  status: 'OPEN' | 'RESOLVED' | 'SUPPRESSED';
  remediation: string | null;
  first_detected: string;
  last_detected: string;
  resolved_at: string | null;
  created_at: string;
  rule_code?: string;
  service?: string;
  resource_name?: string;
  resource_identifier?: string;
  evidence?: Record<string, any>;
  references?: string[];
}

export interface FindingListResponse {
  items: Finding[];
  total: number;
  page: number;
  limit: number;
}
